Show function signatures

sealer3 opened the pull request:
This PR adds a FunctionSignature component that shows a colored function signature and modifies the ParamDeclaration component to support full signatures in addition to short ones with abbreviated tuples.

Here is how it looks. I've added the help button here so that it appears in traces but of course we can remove that part if desired.
![image](https://github.com/otterscan/otterscan/assets/125761775/cd50a1a8-928b-4e16-8fe3-e2879dbc822e)

Something that might improve the look is to switch from `font-code` to `font-mono` for the parameters. I kept it at `font-code` since that is what `ParamDeclaration` uses, but I think `font-mono` would look better:
![image](https://github.com/otterscan/otterscan/assets/125761775/9109c8a2-e071-466f-ba05-f49543415e86)

This probably varies per machine but at least for my default font it looks better :)

This is what it looks like (with font-mono) using 4bytes instead of a verified contract:
![image](https://github.com/otterscan/otterscan/assets/125761775/b0c05557-bf2d-495b-84f7-60f6680eda35)

My intention was to give the decoded function signature a style similar to the decoded log signature which is present in the Logs tab:
![image](https://github.com/otterscan/otterscan/assets/125761775/e2d7d0d4-43e5-4da1-b8fb-6f639318b135)

Closes #1674
Closes #700
Closes #721

wmitsuda commented:
ok, I think you convinced me, I like it.

but we need to improve the UI, when there is sig + docs + input it becomes too noisy, see the screenshot below.

I think we can add some vertical space before the sig (1 or 2 pixels?).

<img width="1032" alt="image" src="https://github.com/otterscan/otterscan/assets/28685/8ce55b5f-f603-4dd5-8f22-cbba5f604efb">

and now I feel like the docs section should have some visual indication that it "belongs" to the sig, what about we try to add a very light yellow bgcolor to the docs box + a left border, like when you add a quote in notion (see a quick and dirty mockup):

<img width="282" alt="image" src="https://github.com/otterscan/otterscan/assets/28685/36f1a354-00d9-421e-8821-cdee0e46c689">

ah, and rebase please :)

sealer3 commented:
I've added a bit more vertical spacing around the signature.

For the docs background color, amber looks better than yellow in my opinion (`bg-amber-100`):
![image](https://github.com/otterscan/otterscan/assets/125761775/7720f916-8561-48bd-a583-063492243355)

Here's one with a light gray background (`bg-gray-100`):
![image](https://github.com/otterscan/otterscan/assets/125761775/412086cd-124b-4cc3-8eee-2a77181483ef)

And no background (this looks best to me):
![image](https://github.com/otterscan/otterscan/assets/125761775/05c6469b-c306-48d9-9065-58ceb1dcae7b)

Once we get the design settled, I'll rebase. I've added a bit more vertical spacing around the signature.

For the docs background color, amber looks better than yellow in my opinion (`bg-amber-100`):
![image](https://github.com/otterscan/otterscan/assets/125761775/7720f916-8561-48bd-a583-063492243355)

Here's one with a light gray background (`bg-gray-100`):
![image](https://github.com/otterscan/otterscan/assets/125761775/412086cd-124b-4cc3-8eee-2a77181483ef)

And no background (this looks best to me):
![image](https://github.com/otterscan/otterscan/assets/125761775/05c6469b-c306-48d9-9065-58ceb1dcae7b)

Once we get the design settled, I'll rebase. Also, do you want to keep the function docs in the transaction action section or remove it in favor of just one place to view the docs, at this function signature?

sealer3 commented:
Here's with a bit of vertical padding added to the block quote (and with the mono font for the parameters, which I like better as it appears on my machine):
![image](https://github.com/otterscan/otterscan/assets/125761775/25ff581e-cd28-4174-b8bd-1f94c01824d5)

wmitsuda commented:
> Here's with a bit of vertical padding added to the block quote (and with the mono font for the parameters, which I like better as it appears on my machine): ![image](https://private-user-images.githubusercontent.com/125761775/297954236-25ff581e-cd28-4174-b8bd-1f94c01824d5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU3MDcwNzcsIm5iZiI6MTcwNTcwNjc3NywicGF0aCI6Ii8xMjU3NjE3NzUvMjk3OTU0MjM2LTI1ZmY1ODFlLWNkMjgtNDE3NC1iOGJkLTFmOTRjMDE4MjRkNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExOVQyMzI2MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lOWQzZmIzZmMzYzgyMWUxMDlkY2ExMzdkMzcyZmMyOTIzOTgxZTM4MjA1NjNmNWUwMjg3ZWUxNmQ5YWViZTEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4-6jUogtvyZPEoKSy2jO_338zwNVKCLQkxOEWEeohQ0)

I think the vertical space + no background looks better.

wmitsuda commented:
about the font, let's keep the -code style, but I'm curious about mono vs fira code. maybe we can improve the -code style font for everyone, can we move that to another issue? (why it looks better, etc, etc)

sealer3 commented:
@wmitsuda Ready for review, with vertical space + no background in the block quote + keeping `font-code`.

wmitsuda reviewed (approved):
lgtm!

wmitsuda commented:
it is cool that not only it looks cool in the tx details page, but we also got that improvement in our ttrace view "for free"

<img width="991" alt="image" src="https://github.com/otterscan/otterscan/assets/28685/7fb32584-2c02-4745-820a-8be17603b4d0">
